### PR TITLE
[Spark-21542][ML][Python]Python persistence helper functions

### PR DESCRIPTION
**Pull request by ajaysaini725**
## What changes were proposed in this pull request?

Added DefaultParamsWriteable, DefaultParamsReadable, DefaultParamsWriter, and DefaultParamsReader to Python to support Python-only persistence of Json-serializable parameters. 

## How was this patch tested?

Instantiated an estimator with Json-serializable parameters (ex. LogisticRegression), saved it using the added helper functions, and loaded it back, and compared it to the original instance to make sure it is the same. This test was both done in the Python REPL and implemented in the unit tests.

Note to reviewers: there are a few excess comments that I left in the code for clarity but will remove before the code is merged to master.

### CODE REVIEW

**Comment by ajaysaini725** [comment]
@jkbradley @thunterdb @MrBago Could you please review this PR?

File: python/pyspark/ml/param/__init__.py [hunk 1]
```diff
@@ -375,6 +375,18 @@ def copy(self, extra=None):
         that._defaultParamMap = {}
         return self._copyValues(that, extra)
 
+    def set(self, param, value):
+        """
```
Comment (WeichenXu123) [inline python/pyspark/ml/param/__init__.py]: We already have `def _set(self, **kwargs)` so does it really need this method ?

File: python/pyspark/ml/util.py [hunk 2]
```diff
@@ -283,3 +289,124 @@ def numFeatures(self):
...
+            for p in paramMap:
+                jsonParams[p.name] = paramMap[p]
+        else:
+            for p in params:
+                jsonParams[p.name] = params[p]
+        basicMetadata = {"class": cls, "timestamp": int(round(time.time() * 1000)),
+                         "sparkVersion": sc.version, "uid": uid, "paramMap": jsonParams}
+        if extraMetadata is not None:
+            basicMetadata.update(extraMetadata)
+        return json.dumps(basicMetadata)
```
Comment (WeichenXu123) [inline python/pyspark/ml/util.py]: I suggest use
`json.dumps(basicMetadata, separators=(',',':'))`
so that output string will compact blank character.
(scala side also do the compaction)

File: python/pyspark/ml/util.py [hunk 3]
```diff
@@ -283,3 +289,124 @@ def numFeatures(self):
...
+        params = instance.extractParamMap()
+        jsonParams = {}
+        if paramMap is not None:
+            for p in paramMap:
+                jsonParams[p.name] = paramMap[p]
+        else:
+            for p in params:
+                jsonParams[p.name] = params[p]
+        basicMetadata = {"class": cls, "timestamp": int(round(time.time() * 1000)),
+                         "sparkVersion": sc.version, "uid": uid, "paramMap": jsonParams}
```
Comment (WeichenXu123) [inline python/pyspark/ml/util.py]: Maybe we should use `long(round(time.time() * 1000))` ?

File: python/pyspark/ml/util.py [hunk 4]
```diff
@@ -283,3 +289,124 @@ def numFeatures(self):
...
+        self.instance = instance
+        self.sc = SparkContext._active_spark_context
+
+    # if a model extends DefaultParamsWriteable this save() function is called
+    def save(self, path):
+        if self.shouldOverwrite:
+            # This command removes a file. Is this enough?
+            os.remove(path)
+        DefaultParamsWriter.save_metadata(self.instance, path, self.sc)
+
```
Comment (WeichenXu123) [inline python/pyspark/ml/util.py]: We need add a `saveImpl` method here and let user to override `saveImpl`, and in `save` function call the `saveImpl` (user should not override `save` method)
like:
```
def save(self, path):
    """
    save metadata first.
    """
    ....
    saveImpl(path)

def saveImpl(self, path):
    raise NotImplementedError()
```

File: python/pyspark/ml/util.py [hunk 5]
```diff
@@ -283,3 +289,124 @@ def numFeatures(self):
...
+    def __init__(self, instance):
+        super(DefaultParamsWriter, self).__init__()
+        self.instance = instance
+        self.sc = SparkContext._active_spark_context
+
+    # if a model extends DefaultParamsWriteable this save() function is called
+    def save(self, path):
+        if self.shouldOverwrite:
+            # This command removes a file. Is this enough?
+            os.remove(path)
```
Comment (WeichenXu123) [inline python/pyspark/ml/util.py]: We use `saveAsTextFile` to save metadata, it is possible to save to `HDFS` or other non-local FS, so here we cannot use `os.remove` to remove it. We need first get the File System of the `path` and call relative API.

File: python/pyspark/ml/util.py [hunk 6]
```diff
@@ -283,3 +289,124 @@ def numFeatures(self):
...
+        return DefaultParamsWriter(self)
+
+
+@inherit_doc
+class DefaultParamsWriter(MLWriter):
+
+    def __init__(self, instance):
+        super(DefaultParamsWriter, self).__init__()
+        self.instance = instance
+        self.sc = SparkContext._active_spark_context
```
Comment (WeichenXu123) [inline python/pyspark/ml/util.py]: Can we move the `sc` into `BaseReadWrite` abstract hierarchy? like the thing scala-side do.

File: python/pyspark/ml/util.py (hunk 4 above)
Comment (ajaysaini725) [inline python/pyspark/ml/util.py]: This implementation exists on the Scala side. However, in Python MLWriter does not even define the saveImpl() function and instead says that the user should override the save() function. I'm not entirely sure which is best to do in Python but as of now everything in pyspark overrides save() instead of saveImpl() so it is probably best to just stick with that pattern.

Comment (WeichenXu123) [inline python/pyspark/ml/util.py]: Good question, in my opinion, the best way is:
when python side use `JavaMLWriter`, just directly override `save`,
but when python side is a custom `MLWriter` implemented in python, tell user to inherit `DefaultParamsWriter` and overwrite `saveImpl`. It will make the interface easier to use.
cc @jkbradley what do you think about this?

File: python/pyspark/ml/param/__init__.py (hunk 1 above)
Comment (jkbradley) [inline python/pyspark/ml/param/__init__.py]: This does look extraneous.  I'm OK with adding it to match the Scala API, but it'd be better to put it in a separate PR and JIRA.

File: python/pyspark/ml/util.py [hunk 7]
```diff
@@ -61,32 +66,82 @@ def _randomUID(cls):
...
+class BaseReadWrite(object):
+    """
+    Base class for MLWriter and MLReader. Stores information about the SparkContext
+    and SparkSession.
+
+    .. versionadded:: 2.3.0
+    """
+
+    def __init__(self):
+        self.sparkSession = None
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: Add leading underscore _sparkSession to be Pythonic

File: python/pyspark/ml/util.py [hunk 8]
```diff
@@ -61,32 +66,82 @@ def _randomUID(cls):
...
+
+    def context(self, sqlContext):
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: No need for this, right?

File: python/pyspark/ml/util.py [hunk 9]
```diff
@@ -61,32 +66,82 @@ def _randomUID(cls):
...
+        Sets the SQL context to use for saving.
+
+        .. note:: Deprecated in 2.1 and will be removed in 3.0, use session instead.
+        """
+        # raise NotImplementedError("MLWriter is not yet implemented for type: %s" % type(self))
+        self.sparkSession = sqlContext.sparkSession
+        return self
+
+    def session(self, sparkSession):
+        """Sets the Spark Session to use for saving."""
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: Copy doc from Scala; no need to rewrite it

File: python/pyspark/ml/util.py [hunk 10]
```diff
@@ -61,32 +66,82 @@ def _randomUID(cls):
...
+        # raise NotImplementedError("MLWriter is not yet implemented for type: %s" % type(self))
+        self.sparkSession = sparkSession
+        return self
+
+    def getOrCreateSparkSession(self):
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: Rename to sparkSession to match Scala API

File: python/pyspark/ml/util.py [hunk 11]
```diff
@@ -61,32 +66,82 @@ def _randomUID(cls):
...
+                os.remove(path)
+            self.saveImpl(path)
+
+    def saveImpl(self, path):
         raise NotImplementedError("MLWriter is not yet implemented for type: %s" % type(self))
 
     def overwrite(self):
         """Overwrites if the output path already exists."""
-        raise NotImplementedError("MLWriter is not yet implemented for type: %s" % type(self))
+        self.shouldOverwrite = True
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: return self (to allow builder pattern); always match Scala API where possible & where it makes sense

File: python/pyspark/ml/util.py [hunk 12]
```diff
@@ -61,32 +66,82 @@ def _randomUID(cls):
...
     def save(self, path):
         """Save the ML instance to the input path."""
+        if isinstance(self, JavaMLWriter):
+            self.saveImpl(path)
+        else:
+            if self.shouldOverwrite:
+                os.remove(path)
+            self.saveImpl(path)
+
+    def saveImpl(self, path):
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: add leading underscore to indicate private/protected

**Review by jkbradley — COMMENTED** [review]
I made a quick pass but am not done yet.  Feel free to update now though.  Thanks!

File: python/pyspark/ml/util.py (hunk 8 above)
Comment (ajaysaini725) [inline python/pyspark/ml/util.py]: Yeah it isn't used I just included the function definition because JavaMLWriter which inherits from MLWriter which inherits from BaseReadWrite uses it. I changed the function body to give a NotImplementedError though.

File: python/pyspark/ml/param/__init__.py (hunk 1 above)
Comment (ajaysaini725) [inline python/pyspark/ml/param/__init__.py]: I included this function because the getAndSetParams function in DefaultParamsReader required a way to set a parameter for an instance given the param object and the value of the param. There wasn't any helper functionality in the Params class that did this so I added the set() function from Scala. Is it okay to leave it for this purpose or is there a better way of handling this case?

File: python/pyspark/ml/param/__init__.py [hunk 13]
```diff
@@ -375,6 +375,18 @@ def copy(self, extra=None):
...
+            Sets the value for a parameter in the parameter map.
```
Comment (jkbradley) [inline python/pyspark/ml/param/__init__.py]: style: change indent -4 spaces

File: python/pyspark/ml/param/__init__.py [hunk 14]
```diff
@@ -375,6 +375,18 @@ def copy(self, extra=None):
...
+        """
+        if not self.hasParam(param.name):
```
Comment (jkbradley) [inline python/pyspark/ml/param/__init__.py]: Reuse _shouldOwn for validation here.

File: python/pyspark/ml/param/__init__.py (hunk 13 above)
Comment (jkbradley) [inline python/pyspark/ml/param/__init__.py]: Also, just copy the doc from Scala.

File: python/pyspark/ml/param/__init__.py (hunk 1 above)
Comment (jkbradley) [inline python/pyspark/ml/param/__init__.py]: OK, makes sense, let's leave it.

File: python/pyspark/ml/param/__init__.py [hunk 15]
```diff
@@ -375,6 +375,18 @@ def copy(self, extra=None):
...
+            raise TypeError('Invalid param %s given for instance %r. %s' % (p.name, self, e))
```
Comment (jkbradley) [inline python/pyspark/ml/param/__init__.py]: Can you add a test which tries to set this with an invalid Param?  This line will not work since e is not defined, but tests pass since this line of code is not tested.

File: python/pyspark/ml/util.py (hunk 4 above)
Comment (jkbradley) [inline python/pyspark/ml/util.py]: I'd strongly prefer to mimic the Scala APIs.

File: mllib/src/main/scala/org/apache/spark/ml/util/ReadWrite.scala [hunk 16]
```diff
@@ -471,3 +471,26 @@ private[ml] object MetaAlgorithmReadWrite {
     List((instance.uid, instance)) ++ subStageMaps
   }
 }
+
+class FileSystemOverwrite extends BaseReadWrite with Logging {
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/util/ReadWrite.scala]: Make this package private: ```private[util]```

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/util/ReadWrite.scala]: Also, I'd make this a static object, eliminate the BaseReadWrite dependency, and just take ```sc``` as an argument.

**Review by jkbradley — COMMENTED** [review]
Some more comments...but not done yet

File: python/pyspark/ml/util.py [hunk 17]
```diff
@@ -86,7 +145,7 @@ def context(self, sqlContext):
 
     def session(self, sparkSession):
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: You can just remove this now, right?  It will be inherited.

File: python/pyspark/ml/util.py [hunk 18]
```diff
@@ -237,6 +300,13 @@ def _load_java_obj(cls, clazz):
             java_obj = getattr(java_obj, name)
         return java_obj
 
+    @classmethod
+    def _load_given_name(cls, java_class):
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: This can be static since you don't use ```cls```

File: python/pyspark/ml/util.py [hunk 19]
```diff
@@ -283,3 +353,143 @@ def numFeatures(self):
         Returns the number of features the model was trained on. If unknown, returns -1
         """
         return self._call_java("numFeatures")
+
+
+@inherit_doc
+class DefaultParamsWritable(MLWritable):
+    """
+    Class for making simple Params types writable. Assumes that all parameters
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: Could you please merge the Scala doc into this to make this more detailed?

File: python/pyspark/ml/util.py [hunk 20]
```diff
@@ -283,3 +353,143 @@ def numFeatures(self):
...
+                         "sparkVersion": sc.version, "uid": uid, "paramMap": jsonParams}
+        if extraMetadata is not None:
+            basicMetadata.update(extraMetadata)
+        return json.dumps(basicMetadata, separators=[',',  ':'])
+
+
+@inherit_doc
+class DefaultParamsReadable(MLReadable):
+    """
+    Class for making simple Params types readable. Assumes that all parameters
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: Could you please merge the Scala doc into this to make this more detailed?

File: python/pyspark/ml/util.py (hunk 18 above)
Comment (jkbradley) [inline python/pyspark/ml/util.py]: Please also stick with camelCase for methods since that's what pyspark does.  (here and elsewhere)

File: python/pyspark/ml/util.py [hunk 21]
```diff
@@ -283,3 +353,143 @@ def numFeatures(self):
...
+        metadataPath = os.path.join(path, "metadata")
+        metadataJson = DefaultParamsWriter.get_metadata_to_save(instance,
+                                                                metadataPath,
+                                                                sc,
+                                                                extraMetadata,
+                                                                paramMap)
+        sc.parallelize([metadataJson], 1).saveAsTextFile(metadataPath)
+
+    @staticmethod
+    def get_metadata_to_save(instance, path, sc, extraMetadata=None, paramMap=None):
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: path is not used

File: python/pyspark/ml/util.py [hunk 22]
```diff
@@ -283,3 +353,143 @@ def numFeatures(self):
...
+
+    def __init__(self, instance):
+        super(DefaultParamsWriter, self).__init__()
+        self.instance = instance
+
+    def saveImpl(self, path):
+        DefaultParamsWriter.save_metadata(self.instance, path, self.sc)
+
+    @staticmethod
+    def save_metadata(instance, path, sc, extraMetadata=None, paramMap=None):
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: add leading underscore to make this private

File: python/pyspark/ml/util.py (hunk 21 above)
Comment (jkbradley) [inline python/pyspark/ml/util.py]: add leading underscore to make this private

File: python/pyspark/ml/util.py [hunk 23]
```diff
@@ -283,3 +353,143 @@ def numFeatures(self):
...
+    def load(self, path):
+        metadata = DefaultParamsReader.loadMetadata(path, self.sc)
+        py_type = DefaultParamsReader.__get_class(metadata['class'])
+        instance = py_type()
+        instance._resetUid(metadata['uid'])
+        DefaultParamsReader.getAndSetParams(instance, metadata)
+        return instance
+
+    @staticmethod
+    def loadMetadata(path, sc, expectedClassName=""):
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: These static methods can be private (add leading underscores)

**Review by jkbradley — COMMENTED** [review]
OK done with my review!  Mostly small cleanups

File: python/pyspark/ml/util.py [hunk 24]
```diff
@@ -61,32 +66,89 @@ def _randomUID(cls):
...
+        method to implement the actual saving of the instance.
         """
         raise NotImplementedError("MLWriter is not yet implemented for type: %s" % type(self))
 
+    def overwrite(self):
+        """Overwrites if the output path already exists."""
+        self.shouldOverwrite = True
+        return self
+
     def session(self, sparkSession):
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: You can remove this instance of session since it is inherited.

File: python/pyspark/ml/util.py [hunk 25]
```diff
@@ -156,28 +218,23 @@ def write(self):
...
 
-    def context(self, sqlContext):
-        """
-        Sets the SQL context to use for loading.
-
-        .. note:: Deprecated in 2.1 and will be removed in 3.0, use session instead.
-        """
-        raise NotImplementedError("MLReader is not yet implemented for type: %s" % type(self))
-
     def session(self, sparkSession):
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: You can remove this instance of session since it is inherited.

File: python/pyspark/ml/util.py [hunk 26]
```diff
@@ -283,3 +341,198 @@ def numFeatures(self):
...
+        return json.dumps(basicMetadata, separators=[',',  ':'])
+
+
+@inherit_doc
+class DefaultParamsReadable(MLReadable):
+    """
+    .. note:: DeveloperApi
+
+    Helper trait for making simple `Params` types readable.  If a `Params` class stores
+    all data as [[pyspark.ml.param.Param]] values, then extending this trait will provide
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: For python docs, follow other examples such as ```:py:class:`MLWriter` ```

File: python/pyspark/ml/util.py [hunk 27]
```diff
@@ -283,3 +341,198 @@ def numFeatures(self):
...
+
+    def __init__(self, instance):
+        super(DefaultParamsWriter, self).__init__()
+        self.instance = instance
+
+    def saveImpl(self, path):
+        DefaultParamsWriter.save_metadata(self.instance, path, self.sc)
+
+    @staticmethod
+    def save_metadata(instance, path, sc, extraMetadata=None, paramMap=None):
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: If this is going to be public, let's use camelCase to match style

**Review by jkbradley — COMMENTED** [review]
Almost done!

File: python/pyspark/ml/util.py [hunk 28]
```diff
@@ -283,3 +333,204 @@ def numFeatures(self):
...
         return self._call_java("numFeatures")
+
+
+@inherit_doc
+class DefaultParamsWritable(MLWritable):
+    """
+    .. note:: DeveloperApi
+
+    Helper trait for making simple `Params` types writable.  If a `Params` class stores
+    all data as [[pyspark.ml.param.Param]] values, then extending this trait will provide
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: There are still some instances of Scaladoc-style class references using double brackets like this: ```[[MyClass]]```.  Switch to the python style like ```:py:class:`MyClass` ```

File: python/pyspark/ml/util.py [hunk 29]
```diff
@@ -283,3 +333,204 @@ def numFeatures(self):
...
+        instance._resetUid(metadata['uid'])
+        DefaultParamsReader.getAndSetParams(instance, metadata)
+        return instance
+
+    @staticmethod
+    def loadMetadata(path, sc, expectedClassName=""):
+        """
+        Load metadata saved using [[DefaultParamsWriter.saveMetadata()]]
+
+        @param expectedClassName  If non empty, this is checked against the loaded metadata.
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: I missed this before: the Scala-style ```@param``` and ```@see``` do not work in Python.  Look in other file like pipeline.py for examples of ```:param```, and maybe use ```.. note::``` to replace ```@see```

**Review by jkbradley — COMMENTED** [review]
almost done..

File: python/pyspark/ml/util.py [hunk 30]
```diff
@@ -283,3 +333,201 @@ def numFeatures(self):
...
         """
         return self._call_java("numFeatures")
+
+
+@inherit_doc
+class DefaultParamsWritable(MLWritable):
+    """
+    .. note:: DeveloperApi
+
+    Helper trait for making simple :py:class`Params` types writable.  If a :py:class`Params`
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: Does this work with the missing colon?  Please make sure to generate the API docs and check them.  See https://github.com/apache/spark/tree/master/docs for instructions

File: python/pyspark/ml/util.py [hunk 31]
```diff
@@ -283,3 +333,201 @@ def numFeatures(self):
...
+    class stores all data as :py:class:'Param' values, then extending this trait will provide
```
Comment (jkbradley) [inline python/pyspark/ml/util.py]: Does normal tick ' work, or do you have to use backticks ` ?

File: python/pyspark/ml/tests.py [hunk 32]
```diff
@@ -1158,6 +1165,33 @@ def test_decisiontree_regressor(self):
...
+    def test_default_read_write(self):
+        temp_path = tempfile.mkdtemp()
+
+        lr = LogisticRegression()
+        lr.setMaxIter(50)
+        lr.setThreshold(.75)
+        writer = DefaultParamsWriter(lr)
+
+        savePath = temp_path + "/lr"
+        writer.saveImpl(savePath)
```
Comment (WeichenXu123) [inline python/pyspark/ml/tests.py]: Why directly call `saveImpl` here ?

File: python/pyspark/ml/util.py [hunk 33]
```diff
@@ -61,33 +66,86 @@ def _randomUID(cls):
...
+    def __init__(self):
+        super(MLWriter, self).__init__()
+        self.shouldOverwrite = False
+
+    def _handleOverwrite(self, path):
+        from pyspark.ml.wrapper import JavaWrapper
+
+        _java_obj = JavaWrapper._new_java_obj("org.apache.spark.ml.util.FileSystemOverwrite")
+        wrapper = JavaWrapper(_java_obj)
+        wrapper._call_java("handleOverwrite", path, True, self.sc._jsc.sc())
```
Comment (WeichenXu123) [inline python/pyspark/ml/util.py]: The wrapper `FileSystemOverwrite`, can we abstract a more generic interface, such as `FileSystem` ? (Maybe it can be used in other place of pyspark, not only "file overwrite", also including other file system operations)

File: python/pyspark/ml/tests.py (hunk 32 above)
Comment (jkbradley) [inline python/pyspark/ml/tests.py]: Good point; +1 for calling save

Comment (ajaysaini725) [inline python/pyspark/ml/tests.py]: Pushed a fix

File: python/pyspark/ml/util.py (hunk 33 above)
Comment (jkbradley) [inline python/pyspark/ml/util.py]: Let's do that when needed.

**Comment by jkbradley** [comment]
LGTM pending tests

**Review by WeichenXu123 — APPROVED** [review]

**Comment by jkbradley** [comment]
Merging with master
Thanks @ajaysaini725 and @WeichenXu123 !
